src/systems/filecoin_markets/storage_market/storage_market_actor_state.go

```diff
@@ -0,0 +1,288 @@
+	p := deal.Proposal()
+
+	st._lockBalance(rt, p.Client(), p.ClientBalanceRequirement())
+	st._lockBalance(rt, p.Provider(), p.ProviderBalanceRequirement())
+}
+
+func (st *StorageMarketActorState_I) _getOnchainDeal(rt Runtime, dealID deal.DealID) deal.OnchainDeal {
+	deal, found := st.Deals()[dealID]
+	if !found {
+		rt.Abort("sm._getOnchainDeal: dealID not found in Deals.")
+	}
+
+	return deal
+}
+
+func (st *StorageMarketActorState_I) _assertPublishedDealState(rt Runtime, dealID deal.DealID) {
+
+	deal := st._getOnchainDeal(rt, dealID)
+
+	if deal.LastPaymentEpoch() != block.ChainEpoch(0) {
+		rt.Abort("sma._assertPublishedDealState: deal is not in PublishedDealState.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertActiveDealState(rt Runtime, dealID deal.DealID) {
+
+	deal := st._getOnchainDeal(rt, dealID)
+
+	if deal.LastPaymentEpoch() <= block.ChainEpoch(0) {
+		rt.Abort("sma._assertActiveDealState: deal is not in ActiveDealState.")
+	}
+}
+
+func (st *StorageMarketActorState_I) _getParticipantBalance(rt Runtime, participant addr.Address) StorageParticipantBalance {
+	balance, found := st.Balances()[participant]
+
+	if !found {
+		rt.Abort("sma._getParticipantBalance: participant balance not found.")
+	}
+
+	return balance
+}
+
+func (st *StorageMarketActorState_I) _getStorageFeeSinceLastPayment(rt Runtime, deal deal.OnchainDeal, newPaymentEpoch block.ChainEpoch) actor.TokenAmount {
+
+	duration := newPaymentEpoch - deal.LastPaymentEpoch()
+	dealP := deal.Deal().Proposal()
+	fee := actor.TokenAmount(0)
+
+	if duration > 0 {
+		unitPrice := dealP.StoragePricePerEpoch()
+		fee := actor.TokenAmount(uint64(duration) * uint64(unitPrice))
+
+		clientBalance := st._getParticipantBalance(rt, dealP.Client())
+
+		if fee > clientBalance.Locked() {
+			rt.Abort("sma._getStorageFeeSinceLastPayment: fee cannot exceed client LockedBalance.")
+		}
+
+	} else {
+		rt.Abort("sma._getStorageFeeSinceLastPayment: no new payment since last payment.")
+	}
+
+	return fee
+
+}
+
+func (st *StorageMarketActorState_I) _slashDealCollateral(rt Runtime, dealP deal.StorageDealProposal) {
+	amountToSlash := dealP.ProviderBalanceRequirement()
+
+	providerBal, found := st.Balances()[dealP.Provider()]
+	if !found {
+		rt.Abort("sma._slashDealCollateral: provider not found in balances.")
+	}
+
+	if providerBal.Locked() < amountToSlash {
+		amountToSlash = providerBal.Locked()
+		// TODO: decide on error handling here
+		panic("TODO")
+	}
+
+	st.Balances()[dealP.Provider()].Impl().Locked_ -= amountToSlash
+	st.DealCollateralSlashed_ += amountToSlash
+
+}
+
+func (st *StorageMarketActorState_I) _terminateDeal(rt Runtime, dealID deal.DealID) {
+
+	deal := st._getOnchainDeal(rt, dealID)
+	st._assertActiveDealState(rt, dealID)
+
+	dealP := deal.Deal().Proposal()
+	delete(st.Deals(), dealID)
+
+	// return client collateral and locked storage fee
+	clientCollateral := dealP.TotalClientCollateral()
+	lockedFee := st._getStorageFeeSinceLastPayment(rt, deal, dealP.EndEpoch())
```

Comment:
no. you want to return full fee.

Reply:
this locked fee is the full fee that has not been unlocked. Not sure if it is possible to pay out full fee, unless we are talking about the same thing here.

Reply:
I am saying we should explicitly ensure the lockedFee is the fullFee in the beginning: specifically, the client is not disallowed (I think) from locking more than what the contract entails. This could lead to fees getting stuck here. 
Two options:
1) check and ensure they only lock up full payment (leading to what you say)
2) let them lock up anything, but return full remaining amount (which would be more than fee to end)

Reply:
I just checked, lockedFee is already the full fee in the beginning. Please let me know if this addresses the issue identified.

https://github.com/filecoin-project/specs/blob/master/src/systems/filecoin_markets/storage_market/storage_market_actor_state.go#L195-L196